### PR TITLE
rsx: Enable multisampling

### PR DESCRIPTION
**Pull request by kd-11**
- Adds MSAA support (vulkan only for now since almost all testers will run vulkan)
- Completely refactors out D3D12 as it was holding back progress with shared code.
- Minor improvement to fp decompiler input modifiers.

### CODE REVIEW

**Comment by Xcedf** [comment]
Resistance 2
AA Auto
E {RSX [0x6c0e4b4]} RSX: Resolve shader:
#version 420
#extension GL_ARB_separate_shader_objects : enable
#extension GL_ARB_shader_stencil_export : enable

layout(std140, set=0, binding=0) uniform static_data{ ivec4 regs[8]; };
layout(set=0, binding=1) uniform sampler2DMS fs0;
layout(set=0, binding=2) uniform usampler2DMS fs1;
layout(pixel_center_integer) in vec4 gl_FragCoord;

void main()
{
	ivec2 out_coord = ivec2(gl_FragCoord.xy);
	ivec2 in_coord = (out_coord / regs[0].xy);
	ivec2 sample_loc = out_coord % ivec2(regs[0].xy);
	int sample_index = sample_loc.x + (sample_loc.y * regs[0].y);
	float frag_depth = texelFetch(fs0, in_coord, sample_index).x;
	uint frag_stencil = texelFetch(fs1, in_coord, sample_index).x;
	gl_FragDepth = frag_depth;
	gl_FragStencilRefARB = int(frag_stencil);
}

F {RSX [0x7075ab4]} RSX: class std::runtime_error thrown: Verification failed (e=0x7f): 
(in file c:\projects\rpcs3\rpcs3\emu\rsx\vk\vkrendertargets.h:334)

AA Disabled 
just

F {RSX [0x7075ab4]} RSX: class std::runtime_error thrown: Verification failed (e=0x7f): 
(in file c:\projects\rpcs3\rpcs3\emu\rsx\vk\vkrendertargets.h:334)

Same for Killzone 2

**Comment by kd-11** [comment]
Resolve shader message is unrelated. I just left it as log_error in case of compilation errors.

**Comment by Xcedf** [comment]
Well it's just verification fail then, and even if i disable it the game still ends with some error

**Comment by kd-11** [comment]
Yea, looks like some files on my end are corrupted; I'm reinstalling some demos and re-running more tests. The verification crash at least is reproducible.

**Comment by arabek** [comment]
Killzone 2 fails for me too, even after nuking all the caches.

**Comment by kd-11** [comment]
Fixed the VkRenderTargets.h crash caused by only testing MSAA x2 during development. MSAA x4 was completely broken which is what killzone 2 is using.

**Comment by kd-11** [comment]
Several other issues identified by testers and under investigation (avoid repeating these):
- Performance slowdown in ni no kuni (even compared to early MSAA-enabled builds)
- Graphics corruption in ni no kuni when MSAA is disabled.

**Comment by Xcedf** [comment]
Now that's what i always expected
Resistance 2 (Native res)
master

**Comment by MarioSonic2987** [comment]
With MSAA enabled, TTT2 freezes in Fallen Garden stage with:
```
·F 0:01:06.184939 {RSX [0x033fdfc]} RSX: class std::runtime_error thrown: Assertion Failed! Vulkan API call failed with unrecoverable error: Device lost (Driver crashed with unspecified error or stopped responding and recovered) (VK_ERROR_DEVICE_LOST)
(in file c:\projects\rpcs3\rpcs3\emu\rsx\vk\vkhelpers.h:978)
```
[Log Broken](https://github.com/RPCS3/rpcs3/files/3262435/RPCS3TTT2.broken.zip)

**Comment by Kravickas** [comment]
Crysis Approve 👍 
msa on
master

**Comment by stride21** [comment]
Master:
Master Ratchet
PR:
ratchet pr
Master:
resis master
PR:
resis 3 pr fixed

File: rpcs3/Emu/RSX/GL/GLRenderTargets.cpp [hunk 1]
```diff
@@ -234,17 +218,12 @@ void GLGSRender::init_buffers(rsx::framebuffer_creation_context context, bool sk
 
 	const u8 color_bpp = get_format_block_size_in_bytes(layout.color_format);
 	const u8 depth_bpp = (layout.depth_format == rsx::surface_depth_format::z16 ? 2 : 4);
```
Comment (scribam) [inline rpcs3/Emu/RSX/GL/GLRenderTargets.cpp]: I think the variable `depth_bpp` can be removed unless you want to replace:
```
m_depth_surface_info.bpp = (layout.depth_format == rsx::surface_depth_format::z16? 2 : 4);
```
by
```
m_depth_surface_info.bpp = depth_bpp;
```

**Comment by MsDarkLow** [comment]
MSAA seems to work as intended in 4. However, I noticed a difference in colors in Yakuza 3, 4 and Ishin, they seem to show up incorrectly.

<details>
<summary>Yakuza 3</summary>
(Doesn't seem to support MSAA)

Master
Y3Master

PR
Y3
</details>

<details>
<summary>Yakuza 4</summary>

Master
Y4Master

PR MSAA
Y4Msaa

PR
Y4
</details>


<details>
<summary>Yakuza Ishin</summary>
(Doesn't seem to support MSAA???)

Master
IshinMaster

PR MSAA
IshinMsaa

PR
Ishin
</details>

**Comment by e1489** [comment]
;=; how do I try this?

**Comment by Artins90** [comment]
> ;=; how do I try this?

Click show all checks below.
Select details near appveryor
Open artifacts and dowload the 17MB file
The antialising option is in the GPU options section, select auto to activate it.

**Comment by raven02** [comment]
D3D12 renderer selection box will be removed as well ?

**Comment by AniLeo** [comment]
d3d12 is not yet fully removed

**Comment by kd-11** [comment]
@MsDarkLow 
> However, I noticed a difference in colors in Yakuza 3, 4 and Ishin, they seem to show up incorrectly.

Maybe show the differences clearly? I cannot spot any differences since I'm not familiar with those games. There are some tweaks to the fragment decompiler based on hardware testing on a PS3, but I still have doubts as to the accuracy; if the commit is wrong compared to PS3 I can revert it from the PR and submit it separately at a later date.

**Comment by anseesaw** [comment]
What does this mean for d3d12，Improve or delete？

**Comment by RainbowCookie32** [comment]
> What does this mean for d3d12，Improve or delete？
> 
> 

DX12 isn't being improved. I'd say that we are one step closer to removal with this PR (I think)

**Comment by MarioSonic2987** [comment]
TTT2 still crashes in Fallen Garden with MSAA enabled:
[Log TTT2 Broken 2](https://github.com/RPCS3/rpcs3/files/3268264/RPCS3.TTT2.Broken.2.gz)

**Comment by MsDarkLow** [comment]
@kd-11 
> Maybe show the differences clearly? I cannot spot any differences since I'm not familiar with those games. There are some tweaks to the fragment decompiler based on hardware testing on a PS3, but I still have doubts as to the accuracy; if the commit is wrong compared to PS3 I can revert it from the PR and submit it separately at a later date.

I got a better scenario with Ishin which is way more clear. Took screenshots in the main menu including how it looks on a PS3.

<details>
<summary>Yakuza Ishin</summary>

PS3
PS3_Like_A_Dragon_Ishin

Master
IshinMaster

PR (both OpenGL & Vulkan look like this)
IshinPR

</details>

**Comment by kd-11** [comment]
@MsDarkLow I changed the decompiler again about an hour ago; need new comparison with the updated code.

**Comment by MsDarkLow** [comment]
Looks relatively the same. 
image

**Comment by kd-11** [comment]
I'll have to check it in a regression tracker; its highly unlikely to get resolved in this PR.

**Comment by Artins90** [comment]
> Looks relatively the same.
Maybe it's hard to notice without a screen with good color reproduction.
In the PS3 and master images there is a yellow glow on the character and the blue of the sky is less dark.
Maybe it has something to do with post processing.

In this other image the top part of the wall on the left side of the screen has a blue tint: https://user-images.githubusercontent.com/9060436/59066374-d3bf7780-88a6-11e9-9270-fad21f8c8e6b.png

**Comment by Margen67** [comment]
Water isn't rendered when anti-aliasing is enabled in Bigduck Sample:
**On**
image
**Off**
image

**Comment by Sandvichero** [comment]
Nier has incorrect colors in this pr.
[LOG](https://github.com/RPCS3/rpcs3/files/3269097/RPCS3.log.gz)

Master:
Nier Master

PR:
Nier PR

**Comment by raveskirza** [comment]
Regression in SC5

Characters are messed up with AA on/off
image


In Master, Characters look fine

**Comment by MarioSonic2987** [comment]
RDR has missing lamps with MSAA enabled.

<details>
 <summary>Images</summary>

**Master**:
image
**PR (MSAA disabled)**:
image
**PR (MSAA enabled)**:
image
</details>

**Comment by AizerMortenort** [comment]
Getting Exception errors 
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/3272810/RPCS3.log)

**Comment by AizerMortenort** [comment]
Desktop Screenshot 2019 06 10 - 19 13 58 39
Here is the error of the crash i get

**Comment by kd-11** [comment]
I have dropped the shader decompiler fixes for now. I'll submit in a separate PR as the issue seems to require more testing. Should fix the graphics glitches in Yakuza and Soul Calibur back to matching master.

**Comment by darkdragonz12** [comment]
soul
Soulcalibur V Fix lastcommit

**Comment by GreyWing3** [comment]
Eternal sonata over-blooming with anti-aliasing on

Screenshot 2019-06-10 20 44 52

**Comment by MarioSonic2987** [comment]
RDR now renders fine without MSAA, but lamps are still missing with MSAA.
<details>
 <summary>RDR regression</summary>

image
image
</details>

Killzone 2 regression is still present (is caused by MSAA on), but renders fine (like master) without MSAA.

<details>
 <summary>Killzone 2</summary>

image
image

</details>

**Comment by kd-11** [comment]
Pushed a few more fixes. 3D and cubemaps (reflections) are still broken in some corner cases; I'm working on that.

**Comment by MarioSonic2987** [comment]
Lamps are now fixed in RDR with MSAA.

image

Killzone 2 regression is now fixed.
image

But Racedriver: Grid still has this "repeated" image with MSAA and no longer reaches title screen without MSAA:
image


[Log Grid Broken (MSAA)](https://github.com/RPCS3/rpcs3/files/3277842/RPCS3.Grid.11-6-19.VK.Broken.zip)

**Comment by Aurez** [comment]
Thanks! Most of my games are working well and better than before :)
But TTT2 is still crashing

**Comment by kd-11** [comment]
@MarioSonic2987 I don't have the repeating image.
image

**Comment by GreyWing3** [comment]
Eternal sonata fixed

**Comment by MarioSonic2987** [comment]
[RenderDoc Grid (Broken)](https://drive.google.com/open?id=1cpOcLU77Cv39rJvO-LBdrFPGOBsdWjDP)

**Comment by kd-11** [comment]
@MarioSonic2987 Retest please.

**Comment by MarioSonic2987** [comment]
Now looks correct with MSAA.
image

**Comment by kd-11** [comment]
@Aurez Confirm if TTT2 crash is fixed.

**Comment by legend800** [comment]
Crowd is missing again on Backbreaker Vengeance. I actually prefer it this way since they're only half rendered on Master, but just letting you know since it's an accuracy thing.

PR:
pr

Master:

**Comment by Margen67** [comment]
Bigduck Sample is now rendering correctly with MSAA enabled.

**Comment by Aurez** [comment]
Just tested again and TTT2 is still crashing 
Image
[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/3282868/RPCS3.zip)

**Comment by Parthorisian** [comment]
SH: Downpour [BLUS30565]:
master - goes ingame(sometimes)/menu
b02
PR - nothing
[SH_Downpour_RPCS3.ZIP](https://github.com/RPCS3/rpcs3/files/3284041/SH_Downpour_RPCS3.ZIP)
b03
WET [BLUS30403]:
master - goes ingame(sometimes)/menu
b11
PR - nothing
[WET_RPCS3.zip](https://github.com/RPCS3/rpcs3/files/3284042/WET_RPCS3.zip)
b05

**Comment by kd-11** [comment]
Those issues seem unrelated. The last one is a crash in main_thread for example. Also, SH Downpour works just fine.
image

**Comment by kd-11** [comment]
@Aurez Install the vulkan SDK from here: https://vulkan.lunarg.com/sdk/home#windows
After that, go to debug settings for the game and enable "debug output" option and run the game. You'll know if it works because your fps should be terrible. Play the game as usual, when it crashes, attach the logfile here. It's the only way to actually debug a DEVICE_LOST error.

**Comment by Parthorisian** [comment]
If Win7 support was discontinued by Nekotekina then definitely yes

**Comment by Leopard1907** [comment]
@Parthorisian How did you come up with that?

**Comment by kd-11** [comment]
This branch diverged from master a while ago, maybe its just a bug that has been fixed in master but the commits are missing from this branch. We'll know after I rebase which I don't want to attempt before fixing the remaining driver crashes.

**Comment by kd-11** [comment]
@Aurez @MarioSonic2987 You're both running GTX 970s; I checked this issue and its not present on pascal or even AMD cards. Does it crash immediately after the shader is printed to the log window?

**Comment by Aurez** [comment]
@kd-11 sorry just back at home now. I will do some tests and report

**Comment by Aurez** [comment]
The crash screen and log I just got with your latest update.
It crashes after approx 40s of gameplay
I will install the sdk and share the debug output log in a few minutes
Imagea
[RPCS3a.zip](https://github.com/RPCS3/rpcs3/files/3287687/RPCS3a.zip)

**Comment by Aurez** [comment]
With sdk installed and debug output log
Imagedebug
[RPCS3_debug.zip](https://github.com/RPCS3/rpcs3/files/3287786/RPCS3_debug.zip)

**Comment by kd-11** [comment]
Oh, this looks like this is caused by heap corruption, not actually bad rendering. I am unable to reproduce this on windows 10 (both 1809 and 1903).

**Comment by kd-11** [comment]
@Aurez Does it crash if you open the stage in practice mode and just stand there?

**Comment by Aurez** [comment]
It crashes in Vs and Practice modes. Sometime quickly and sometime not (I just managed to play 3 fights without any crashes) But most often it crashes soon or late (at least on the 2 last stages, I didn't tried a lot the others ones but they seems to work fine)

**Comment by kd-11** [comment]
TTT2 crash is now fixed.

**Comment by AizerMortenort** [comment]
Hello anyone able to fix Dynasty Warriors Gundam Reborn crash ?

**Comment by AizerMortenort** [comment]
Whats so wrong into bringing the crash from Dynasty Warriors Gundam Reborn to a developer's attention? Game crashes for years with same error so i dont see any update fixing it. Thanks

**Comment by RainbowCookie32** [comment]
Because it's completely unrelated to this PR

**Comment by AniLeo** [comment]
You're wasting a developer's time with a false regression report, it's completely unrelated to this pull request

**Comment by AizerMortenort** [comment]
Still this is not a reason for hate.

**Pull request merged**